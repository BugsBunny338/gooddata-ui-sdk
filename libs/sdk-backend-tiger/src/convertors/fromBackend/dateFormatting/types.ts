// (C) 2020 GoodData Corporation
import { DateAttributeGranularity } from "@gooddata/sdk-model";

export type DateFormatter = (value: Date, granularity: DateAttributeGranularity) => string;
