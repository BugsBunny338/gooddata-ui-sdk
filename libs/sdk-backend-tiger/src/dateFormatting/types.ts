// (C) 2020 GoodData Corporation
import { CatalogDateAttributeGranularity } from "@gooddata/sdk-model";

export type DateFormatter = (value: Date, granularity: CatalogDateAttributeGranularity) => string;
