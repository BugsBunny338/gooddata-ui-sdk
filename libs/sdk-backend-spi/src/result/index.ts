// (C) 2019 GoodData Corporation

import { IExportConfig, IExportResult } from "../export";

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IExecutionResult {
    /**
     * Unique identifier of the execution result.
     */
    readonly id: string;
    readonly dimensions: IResultDimension[];

    readAll(): Promise<IDataView>;

    readView(offset: number[], limit: number[]): Promise<IDataView>;

    export(options: IExportConfig): Promise<IExportResult>;

    equals(other: IExecutionResult): boolean;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IDataView {
    readonly offset: number[];
    readonly limit: number[];
    readonly headerItems?: IResultHeaderItem[][][];
    readonly data: DataValue[][] | DataValue[];
    readonly totals?: DataValue[][][];

    next(...dims: boolean[]): Promise<IDataView | null>;

    pageUp(): Promise<IDataView | null>;

    pageDown(): Promise<IDataView | null>;

    pageLeft(): Promise<IDataView | null>;

    pageRight(): Promise<IDataView | null>;
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export type DataValue = null | string | number;

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IMeasureHeaderItem {
    measureHeaderItem: {
        uri?: string;
        identifier?: string;
        localIdentifier: string;
        name: string;
        format: string;
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface ITotalHeaderItem {
    totalHeaderItem: {
        name: string;
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IMeasureGroupHeader {
    measureGroupHeader: {
        items: IMeasureHeaderItem[];
        totalItems?: ITotalHeaderItem[];
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IAttributeHeader {
    attributeHeader: {
        uri: string;
        identifier: string;
        localIdentifier: string;
        name: string;
        totalItems?: ITotalHeaderItem[];
        formOf: {
            uri: string;
            identifier: string;
            name: string;
        };
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export type IHeader = IMeasureGroupHeader | IAttributeHeader;

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IResultDimension {
    headers: IHeader[];
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IResultAttributeHeaderItem {
    attributeHeaderItem: {
        uri: string;
        name: string;
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IResultMeasureHeaderItem {
    measureHeaderItem: {
        name: string;
        order: number;
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export interface IResultTotalHeaderItem {
    totalHeaderItem: {
        name: string;
        type: string;
    };
}

/**
 * TODO: SDK8: add docs
 * @public
 */
export type IResultHeaderItem =
    | IResultAttributeHeaderItem
    | IResultMeasureHeaderItem
    | IResultTotalHeaderItem;
