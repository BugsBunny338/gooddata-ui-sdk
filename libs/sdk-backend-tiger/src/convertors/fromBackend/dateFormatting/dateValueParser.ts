// (C) 2020 GoodData Corporation
import parse from "date-fns/parse";
import identity from "lodash/identity";
import { DateAttributeGranularity } from "@gooddata/sdk-model";
import { UnexpectedError } from "@gooddata/sdk-backend-spi";

type ValueTransform = (value: string) => string;

const granularityParseValueTransformations: {
    [granularity in DateAttributeGranularity]?: ValueTransform;
} = {
    "GDC.time.day_in_week": (value) => {
        // server returns 0 = Sunday, 6 = Saturday
        // date-fns expects 1 = Monday, 7 = Sunday
        return value === "0" ? "7" : value;
    },
};

const granularityParsePatterns: { [granularity in DateAttributeGranularity]?: string } = {
    "GDC.time.date": "yyyy-MM-dd", // 2020-01-31
    "GDC.time.day_in_month": "d", // 1-31
    "GDC.time.day_in_week": "i", // 1-7
    "GDC.time.day_in_year": "D", // 1-366
    "GDC.time.month": "yyyy-MM", // 2020-06
    "GDC.time.month_in_year": "M", // 1-12
    "GDC.time.quarter": "yyyy-q", // 2020-1
    "GDC.time.quarter_in_year": "q", // 1-4
    "GDC.time.week_us": "YYYY-ww", // 2020-05
    "GDC.time.week_in_year": "w", // 5
    "GDC.time.year": "yyyy", // 2020
};

/**
 * Parses a string representation of a date of a given granularity to a Date object.
 * @param value - value to parse.
 * @param granularity - granularity to assume when parsing the value.
 * @internal
 */
export const parseDateValue = (value: string, granularity: DateAttributeGranularity): Date => {
    const parsePattern = granularityParsePatterns[granularity];
    if (!parsePattern) {
        throw new UnexpectedError(`No date parser for the "${granularity}" granularity available.`);
    }

    const valueTransform = granularityParseValueTransformations[granularity] ?? (identity as ValueTransform);

    return parse(valueTransform(value), parsePattern, new Date(), {
        useAdditionalDayOfYearTokens: true, // for day of year parsing
        useAdditionalWeekYearTokens: true, // for week parsing
    });
};
