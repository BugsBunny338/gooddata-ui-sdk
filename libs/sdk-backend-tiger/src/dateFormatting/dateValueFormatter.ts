// (C) 2020 GoodData Corporation
import format from "date-fns/format";
import parse from "date-fns/parse";
import { Locale } from "date-fns"; // TODO should be import type when we have prettier 2
import { enUS, de, es, fr, ja, nl, pt, ptBR, zhCN } from "date-fns/locale";
import identity = require("lodash/identity");
import { CatalogDateAttributeGranularity } from "@gooddata/sdk-model";
import { UnexpectedError } from "@gooddata/sdk-backend-spi";

type ValueTransform = (value: string) => string;

interface IGranularityFormatterMeta {
    valueTransform?: ValueTransform;
    parsePattern: string;
    formatPattern: string;
}

const granularityFormatterMetas: {
    [granularity in CatalogDateAttributeGranularity]?: IGranularityFormatterMeta;
} = {
    "GDC.time.date": {
        parsePattern: "yyyy-MM-dd", // 2020-01-31
        formatPattern: "P", // 01/31/2020
    },
    "GDC.time.day_in_month": {
        parsePattern: "d", // 1-31
        formatPattern: "dd", // 01-31
    },
    "GDC.time.day_in_week": {
        valueTransform: value => {
            // server returns 0 = Sunday, 6 = Saturday
            // date-fns expects 1 = Monday, 7 = Sunday
            return value === "0" ? "7" : value;
        },
        parsePattern: "i", // 1-7
        formatPattern: "EEEE", // Monday-Sunday
    },
    "GDC.time.day_in_year": {
        parsePattern: "D", // 1-366
        formatPattern: "D", // 1-366
    },
    "GDC.time.month": {
        parsePattern: "yyyy-MM", // 2020-06
        formatPattern: "LLL yyyy", // Jun 2020
    },
    "GDC.time.month_in_year": {
        parsePattern: "M", // 1-12
        formatPattern: "LLL", // Jan-Dec
    },
    "GDC.time.quarter": {
        parsePattern: "yyyy-q", // 2020-1
        formatPattern: "qqq/yyyy", // Q1/2020
    },
    "GDC.time.quarter_in_year": {
        parsePattern: "q", // 1-4
        formatPattern: "qqq", // Q1-Q4
    },
    "GDC.time.week_us": {
        parsePattern: "YYYY-ww", // 2020-05
        formatPattern: "'W'w/YYYY", // W5/2020
    },
    "GDC.time.week_in_year": {
        parsePattern: "w", // 5
        formatPattern: "'W'w", // W5
    },
    "GDC.time.year": {
        parsePattern: "yyyy", // 2020
        formatPattern: "yyyy", // 2020
    },
};

const localeConversions: { [key: string]: Locale } = {
    "en-us": enUS,
    "de-de": de,
    "es-es": es,
    "fr-fr": fr,
    "ja-jp": ja,
    "nl-nl": nl,
    "pt-br": ptBR,
    "pt-pt": pt,
    "zh-hans": zhCN,
};

export type DateValueFormatter = (value: string, granularity: CatalogDateAttributeGranularity) => string;

export function createDateValueFormatter(locale: string): DateValueFormatter {
    const convertedLocale = localeConversions[locale.toLowerCase()];
    if (!convertedLocale) {
        throw new UnexpectedError(
            `Unsupported locale "${locale}". Supported locales are ${Object.keys(localeConversions).join(
                ", ",
            )}`,
        );
    }

    return (value, granularity) => {
        const converterMeta = granularityFormatterMetas[granularity];
        if (!converterMeta) {
            throw new UnexpectedError(`No date formatter for the "${granularity}" granularity available.`);
        }

        const { formatPattern, parsePattern, valueTransform = identity as ValueTransform } = converterMeta;

        const parsed = parse(valueTransform(value), parsePattern, new Date(), {
            useAdditionalDayOfYearTokens: true, // for day of year parsing
            useAdditionalWeekYearTokens: true, // for week parsing
        });

        return format(parsed, formatPattern, {
            locale: convertedLocale,
            useAdditionalDayOfYearTokens: true, // for day of year formatting
            useAdditionalWeekYearTokens: true, // for week formatting
        });
    };
}

export const defaultValueDateFormatter = createDateValueFormatter("en-us");
