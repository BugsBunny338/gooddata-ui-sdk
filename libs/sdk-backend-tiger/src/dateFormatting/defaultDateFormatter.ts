// (C) 2020 GoodData Corporation
import format from "date-fns/format";
import { enUS, de, es, fr, ja, nl, pt, ptBR, zhCN } from "date-fns/locale";
import { CatalogDateAttributeGranularity } from "@gooddata/sdk-model";
import { UnexpectedError } from "@gooddata/sdk-backend-spi";
import { DateFormatter } from "./types";

const granularityFormatPatterns: {
    [granularity in CatalogDateAttributeGranularity]?: string;
} = {
    "GDC.time.date": "P", // 01/31/2020
    "GDC.time.day_in_month": "dd", // 01-31
    "GDC.time.day_in_week": "EEEE", // Monday-Sunday
    "GDC.time.day_in_year": "D", // 1-366
    "GDC.time.month": "LLL yyyy", // Jun 2020
    "GDC.time.month_in_year": "LLL", // Jan-Dec
    "GDC.time.quarter": "qqq/yyyy", // Q1/2020
    "GDC.time.quarter_in_year": "qqq", // Q1-Q4
    "GDC.time.week_us": "'W'w/YYYY", // W5/2020
    "GDC.time.week_in_year": "'W'w", // W5
    "GDC.time.year": "yyyy", // 2020
};

const localeConversions = {
    "en-US": enUS,
    "de-DE": de,
    "es-ES": es,
    "fr-FR": fr,
    "ja-JP": ja,
    "nl-NL": nl,
    "pt-BR": ptBR,
    "pt-PT": pt,
    "zh-Hans": zhCN,
};

/**
 * Creates a default date formatting function for a given locale.
 * @param locale - the code of the locale to use
 * @public
 */
export const createDefaultDateFormatter = (
    locale: keyof typeof localeConversions = "en-US",
): DateFormatter => {
    const convertedLocale = localeConversions[locale];
    if (!convertedLocale) {
        throw new UnexpectedError(
            `Unsupported locale "${locale}". Supported locales are ${Object.keys(localeConversions).join(
                ", ",
            )}`,
        );
    }

    return (value, granularity) => {
        const formatPattern = granularityFormatPatterns[granularity];
        if (!formatPattern) {
            throw new UnexpectedError(`No date formatter for the "${granularity}" granularity available.`);
        }

        return format(value, formatPattern, {
            locale: convertedLocale,
            useAdditionalDayOfYearTokens: true, // for day of year formatting
            useAdditionalWeekYearTokens: true, // for week formatting
        });
    };
};
