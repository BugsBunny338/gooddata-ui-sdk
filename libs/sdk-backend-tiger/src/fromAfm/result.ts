// (C) 2019-2020 GoodData Corporation
import { DataValue, IResultAttributeHeader, IResultHeader } from "@gooddata/sdk-backend-spi";
import { Execution } from "@gooddata/gd-tiger-client";
import isArray = require("lodash/isArray");
import isAttributeHeaderItem = Execution.isAttributeHeaderItem;

export type TransformerResult = {
    readonly headerItems: IResultHeader[][][];
    readonly data: DataValue[][] | DataValue[];
    readonly offset: number[];
    readonly count: number[];
};

function transformHeaderItems(headerItems?: Execution.IResultHeaderItem[][][]): IResultHeader[][][] {
    if (!headerItems) {
        return [[[]]];
    }

    return headerItems.map(dim => {
        return dim.map(dimHeaders => {
            return dimHeaders.map(
                (headerItem): IResultHeader => {
                    const item = headerItem;

                    if (isAttributeHeaderItem(item)) {
                        const newItem: IResultAttributeHeader = {
                            attributeHeaderItem: {
                                uri: `/fake/${item.attributeHeaderItem.name}`,
                                name: item.attributeHeaderItem.name,
                            },
                        };

                        return newItem;
                    }

                    return item;
                },
            );
        });
    });
}

function calculateOffset(data: DataValue[][] | DataValue[]): number[] {
    if (!data) {
        return [0];
    }

    return isArray(data[0]) ? [0, 0] : [0];
}

function calculateCount(data: DataValue[][] | DataValue[]): number[] {
    const xCount = data.length;
    const yCount = isArray(data[0]) ? (data[0] as DataValue[]).length : 0;

    return [xCount, yCount];
}

export function transformExecutionResult(result: Execution.IExecutionResult): TransformerResult {
    return {
        data: result.data,
        headerItems: transformHeaderItems(result.headerItems),
        offset: calculateOffset(result.data),
        count: calculateCount(result.data),
    };
}
