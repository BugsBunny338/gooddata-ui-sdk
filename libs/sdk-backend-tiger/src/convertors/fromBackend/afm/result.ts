// (C) 2019-2020 GoodData Corporation
import {
    DataValue,
    IResultHeader,
    IDimensionDescriptor,
    isAttributeDescriptor,
    IDimensionItemDescriptor,
    IMeasureDescriptor,
    isMeasureGroupDescriptor,
} from "@gooddata/sdk-backend-spi";
import { Execution, AttributeGranularityResourceAttribute } from "@gooddata/api-client-tiger";
import isResultAttributeHeader = Execution.isResultAttributeHeader;
import isResultMeasureHeader = Execution.isResultMeasureHeader;
import { DateAttributeGranularity } from "@gooddata/sdk-model";
import { toSdkGranularity } from "../dateGranularityConversions";
import { DateFormatter } from "../dateFormatting/types";
import { createDateValueFormatter } from "../dateFormatting/dateValueFormatter";

export type TransformerResult = {
    readonly headerItems: IResultHeader[][][];
    readonly data: DataValue[][] | DataValue[];
    readonly offset: number[];
    readonly count: number[];
    readonly total: number[];
};

// gets all the enum values
const supportedSuffixes: string[] = Object.keys(AttributeGranularityResourceAttribute)
    .filter((item) => isNaN(Number(item)))
    .map(
        (key) =>
            AttributeGranularityResourceAttribute[key as keyof typeof AttributeGranularityResourceAttribute],
    );

function getGranularity(header: IDimensionItemDescriptor): DateAttributeGranularity | undefined {
    if (!isAttributeDescriptor(header)) {
        return undefined;
    }

    const { identifier } = header.attributeHeader.formOf;
    const suffix = identifier.substr(identifier.lastIndexOf(".") + 1);

    return supportedSuffixes.includes(suffix)
        ? toSdkGranularity(suffix as AttributeGranularityResourceAttribute)
        : undefined; // not a date attribute
}

function getMeasuresFromDimensions(dimensions: IDimensionDescriptor[]): IMeasureDescriptor[] {
    for (const dim of dimensions) {
        const measureGroup = dim.headers.find(isMeasureGroupDescriptor);

        if (measureGroup) {
            return measureGroup.measureGroupHeader.items;
        }
    }

    return [];
}

function transformHeaderItems(
    dimensions: IDimensionDescriptor[],
    dateFormatter: DateFormatter,
    dimensionHeaders?: Execution.IDimensionHeader[],
): IResultHeader[][][] {
    if (!dimensionHeaders) {
        return [[[]]];
    }

    const measureDescriptors = getMeasuresFromDimensions(dimensions);
    const dateValueFormatter = createDateValueFormatter(dateFormatter);

    return dimensionHeaders.map((dimensionHeader, dimensionIndex) => {
        return dimensionHeader.headerGroups.map((headerGroup, headerGroupIndex) => {
            const granularity = getGranularity(dimensions[dimensionIndex].headers[headerGroupIndex]);
            return headerGroup.headers.map(
                (header): IResultHeader => {
                    if (isResultAttributeHeader(header)) {
                        const primaryLabel =
                            header.attributeHeader.primaryLabelValue ?? header.attributeHeader.labelValue;

                        /*
                         * Funny stuff #1 - we have to set 'uri' to some made-up value resembling the URIs sent by bear. This
                         * is because pivot table relies on the format of URIs. Ideally we would refactor pivot table to
                         * not care about this however this aspect is like a couple of eggs that hold the pivot spaghetti
                         * together - cannot be easily untangled.
                         */
                        return {
                            attributeHeaderItem: {
                                uri: `/obj/${headerGroupIndex}/elements?id=${primaryLabel}`,
                                name: granularity
                                    ? dateValueFormatter(header.attributeHeader.labelValue, granularity)
                                    : header.attributeHeader.labelValue,
                            },
                        };
                    }

                    if (isResultMeasureHeader(header)) {
                        /*
                         * Funny stuff #2 - Tiger sends just the measure index in the measure headers. This is the index of the
                         * measure descriptor within the measure group. The code looks up the measure descriptor so that
                         * it can then fill in the `name` to the one in the descriptor
                         */
                        const measureIndex = header.measureHeader.measureIndex;

                        return {
                            measureHeaderItem: {
                                name: measureDescriptors[measureIndex]?.measureHeaderItem.name,
                                order: measureIndex,
                            },
                        };
                    }

                    return {
                        totalHeaderItem: {
                            name: header.totalHeader.name,
                            type: header.totalHeader.type,
                        },
                    };
                },
            );
        });
    });
}

export function transformExecutionResult(
    result: Execution.IExecutionResult,
    dimensions: IDimensionDescriptor[],
    dateFormatter: DateFormatter,
): TransformerResult {
    return {
        data: result.data,
        headerItems: transformHeaderItems(dimensions, dateFormatter, result.dimensionHeaders),
        offset: result.paging.offset,
        count: result.paging.count,
        total: result.paging.total,
    };
}
