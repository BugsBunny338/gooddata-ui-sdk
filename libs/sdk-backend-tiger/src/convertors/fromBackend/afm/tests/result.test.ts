// (C) 2020 GoodData Corporation
import { transformExecutionResult } from "../result";
import { createDefaultDateFormatter } from "../../dateFormatting/defaultDateFormatter";
import { mockResult, mockDimensions } from "./result.fixture";

describe("transformExecutionResult", () => {
    it("should format date dimensions values", () => {
        const actual = transformExecutionResult(mockResult, mockDimensions, createDefaultDateFormatter());
        expect(actual).toMatchSnapshot();
    });
});
