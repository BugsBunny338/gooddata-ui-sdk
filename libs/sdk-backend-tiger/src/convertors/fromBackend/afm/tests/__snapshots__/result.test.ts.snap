// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformExecutionResult should format date dimensions values 1`] = `
Object {
  "count": Array [
    1,
    2,
  ],
  "data": Array [
    Array [
      "20.0",
      "40.2",
    ],
  ],
  "headerItems": Array [
    Array [
      Array [
        Object {
          "measureHeaderItem": Object {
            "name": "Sum of duration",
            "order": 0,
          },
        },
      ],
    ],
    Array [
      Array [
        Object {
          "attributeHeaderItem": Object {
            "name": "Q4/1906",
            "uri": "/obj/0/elements?id=1906-4",
          },
        },
        Object {
          "attributeHeaderItem": Object {
            "name": "Q1/1910",
            "uri": "/obj/0/elements?id=1910-1",
          },
        },
      ],
    ],
  ],
  "offset": Array [
    0,
    0,
  ],
  "total": Array [
    1,
    2,
  ],
}
`;
